jraft-core/src/main/java/com/alipay/sofa/jraft/core/IteratorImpl.java

```diff
@@ -46,10 +46,12 @@ public class IteratorImpl {
     private final long          firstClosureIndex;
     private long                currentIndex;
     private final long          committedIndex;
-    private long                fsmCommittedIndex;         // fsm commit index
-    private LogEntry            currEntry = new LogEntry(); // blank entry
+    private long                fsmCommittedIndex;          // fsm commit index
+    private LogEntry            currEntry  = new LogEntry(); // blank entry
     private final AtomicLong    applyingIndex;
     private RaftException       error;
+    private boolean             autoCommit = false;         // Default not enabled
```

Comment:
In the past, the default value should be true, otherwise there may be backward compatibility problems

Reply:
Okay, I'll make the changes and push it back up again

Reply:
> In the past, the default value should be true, otherwise there may be backward compatibility problems

So now, as long as the `hasNext` method is called, it defaults to automatic submission once? This means that the automatic submission switch defaults to the on state

Reply:
Yes, I think this is better, the default case should not care about log submission, and business processing throws an exception in the case of certainly can not submit this log, the manual submission option to the user, the framework level should be handled by default. @fengjiachun @killme2008 what do you think?

Reply:
Auto-commit should not be enabled by default to be compatible with old releases.

Reply:
In the past I've always used `next`, doesn't this already help me submit? @killme2008

Reply:
> No, it will not be submitted until returns from `StateMachine#onApply`, and you can manually rollback by `Iterator#setErrorAndRollback`. If the `onApply` throws a runtime exception, this batch will be rollbacked automatically.

StateMachine#onApply运行完毕后就会提交，这个提交就是自动提交，现在默认值改为了false，变成对每个iter的进行提交，默认又是false，假设我某个iter没有提交，也没有抛出异常，此时会发生什么？比如都是i++的动作有10个，我进行到第9个不提交，继续next第10个元素并commit，会发生什么？

Reply:
没明白，默认改为 false，也就是现在的行为了，并没有什么不同。改为 true，就是每条都 commit。

Reply:
哦，我理解了，这个名字可能有点误导，所谓 auto-commit 是说对每一条自动 commit，设置不设置都不影响 `onApply` 后自动 commit 的逻辑。

Perhaps we should rename it to `autoCommitPerLog`.

Reply:
> 



> @1294566108 I think we should call `commit` in `Iterator#next()`. It means committing the state machine before getting the next element, just like we don't have auto-commit mode but we want to commit on every log:
> 
> ```java
> while(it.hasNext()) {
>    byte [] data = it.getData();
>    it.commit();
>    it.next();   
> }
> ```
> 
> It looks like we don't need the variable `lastCommit`, commit many times for the same log index has no side effects.

所以现在修改为：
1.将自动提交的属性名改成`autoCommitPerLog`
2.在用户开启自动提交的前提下，在调用`next`方法时去触发`commit`方法，而不是`hasNext`方法时
```
    @Override
    public ByteBuffer next() {
        // commit the current log if auto-commit mode is on and not yet committed
        if (impl.getAutoCommit() && !impl.hasError()) {
            commit();
        }
        final ByteBuffer data = getData();
        if (hasNext()) {
            this.impl.next();
        }
        return data;
    }
```

Reply:
> 哦，我理解了，这个名字可能有点误导，所谓 auto-commit 是说对每一条自动 commit，设置不设置都不影响 `onApply` 后自动 commit 的逻辑。
> 
> Perhaps we should rename it to `autoCommitPerLog`.

明白了谢谢,确实是我理解错了,看来是别误导到了哈哈,谢谢